File: Java/hellointerview/03-intervals/insertIntervals.java
import java.util;

class Solution {
  public int[][] insertIntervals(int[][] intervals, int[] newInterval) {
    List<int[]> result = new ArrayList<>();

    int i = 0;
    int n = intervals.length;

    while (i < n && intervals[i][0] < newInterval[0]) {
      result.add(intervals[i]);
      i++;
    }

    while (i < n && intervals[i][0] <= newInterval[1]) {
      newInterval[0] = Math.min(intervals[i][0], newInterval[0]);
      newInterval[1] = Math.max(intervals[i][1], newInterval[1]);
      i++;
    }

    result.add(newInterval);

    for (int j = i; j < n; j++) {
      results.add(intervals[j]);
    }

    return result.toArray(new int[result.size()][]);
  }
}

File: TS/hellointerview/intervals/insertIntervals.ts
// 1,3  6,9 (2,5)
class Solution {
  insertIntervals(intervals: number[][], newInterval: number[]): number[][] {
    const merged: number[][] = [];

    let i = 0;
    const n = intervals.length;

    while (i < n && intervals[i][1] < newInterval[0]) {
      merged.push(intervals[i]);
      i++;
    }

    while (i < n && intervals[i][0] <= newInterval[1]) {
      newInterval[0] = Math.min(intervals[i][0], newInterval[0]);
      newInterval[1] = Math.max(intervals[i][1], newInterval[1]);
      i++;
    }

    merged.push(newInterval);

    while (i < n) {
      merged.push(intervals[i]);
      i++;
    }
    return merged;
  }

}
